File: src/main/scala/top/criwits/scaleda/idea/windows/tool/ScaledaToolWindowFactory.scala
package top.criwits.scaleda
package idea.windows.tool

import idea.ScaledaBundle
import idea.utils.{MainLogger, OutputLogger, ProjectNow, invokeLater}
import idea.windows.tool.logging.{ConsoleTabManager, ScaledaLoggingService}
import idea.windows.tool.message.ScaledaMessageTab

import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.wm.{ToolWindow, ToolWindowFactory}

class ScaledaToolWindowFactory extends ToolWindowFactory {
  private var created = false
  override def init(toolWindow: ToolWindow) = {
    super.init(toolWindow)
    MainLogger.info("ScaledaToolWindowFactory#init")
    invokeLater {
      // ProjectNow().foreach(p => createToolWindowContent(p, toolWindow))
    }
  }

  override def createToolWindowContent(
      project: Project,
      toolWindow: ToolWindow
  ): Unit = {
    if (created) return
    else created = true
    MainLogger.info("ScaledaToolWindowFactory#createToolWindowContent")
    val service = project.getService(classOf[ScaledaLoggingService])
    val tabManager =
      new ConsoleTabManager(project, toolWindow.getContentManager)
    Disposer.register(service, tabManager)

    val messageTab = ScaledaMessageTab(project, tabManager = Some(tabManager))

    tabManager.addPanel(messageTab, messageTab.getDisplayName)
    tabManager.addTab(OutputLogger.LOGGER_ID, ScaledaBundle.message("windows.tool.log.output.title"), switchTo = false)
    tabManager.addTab(MainLogger.LOGGER_ID, ScaledaBundle.message("windows.tool.log.main.title"), switchTo = false)
  }
}

object ScaledaToolWindowFactory {
  val WINDOW_ID = "Scaleda Messages"
}

File: src/main/scala/top/scaleda/idea/windows/bottomPanel/ScaledaToolWindowFactory.java
package top.scaleda.idea.windows.bottomPanel;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowFactory;
import org.jetbrains.annotations.NotNull;
import top.scaleda.idea.ScaledaBundle;
import top.scaleda.idea.utils.ScaledaIdeaLogger;
import top.scaleda.idea.windows.bottomPanel.console.ConsoleTabManager;
import top.scaleda.idea.windows.bottomPanel.message.MessageListPanel;
import top.scaleda.idea.windows.bottomPanel.rvcd.RvcdPanel;

public class ScaledaToolWindowFactory implements ToolWindowFactory {
  // @Override
  // public boolean shouldBeAvailable(@NotNull Project project) {
  //   return YmlRootManager.getInstance(project).getRoots().nonEmpty();
  // }

  @Override
  public void createToolWindowContent(@NotNull Project project, @NotNull ToolWindow toolWindow) {
    ConsoleService service = project.getService(ConsoleService.class);
    ConsoleTabManager tabManager = new ConsoleTabManager(project, toolWindow.getContentManager());
    Disposer.register(service, tabManager);
    service.setConsoleTabManager(tabManager);
    MessageListPanel messageTab = new MessageListPanel(project);
    RvcdPanel rvcdPanel = new RvcdPanel(project);
    Disposer.register(tabManager, messageTab);
    Disposer.register(tabManager, rvcdPanel);
    tabManager.addPanel(rvcdPanel, ScaledaBundle.message("windows.tool.log.waveform.title"), false, false);
    tabManager.addPanel(messageTab, ScaledaBundle.message("windows.tool.log.message.title"), false, false);
    tabManager.addConsoleTab(ScaledaIdeaLogger.getId(), ScaledaBundle.message("windows.tool.log.console.title"), false, false);
  }
}
